src/main/java/com/gtnewhorizons/modularui/api/math/MathExpression.java

```diff
@@ -1,315 +1,593 @@
 package com.gtnewhorizons.modularui.api.math;
 
-import java.text.ParseException;
+import java.text.NumberFormat;
+import java.text.ParsePosition;
 import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
+import java.util.function.BiFunction;
+import java.util.regex.Pattern;
 
-import com.gtnewhorizons.modularui.common.widget.textfield.TextFieldWidget;
+import org.jetbrains.annotations.NotNull;
 
 public class MathExpression {
 
-    private static final List<Object> DEFAULT = Collections.singletonList(0);
+    /**
+     * Matches any string that can be evaluated to a number. The pattern might be too generous, i.e., matches some
+     * strings that do not evaluate to a valid value. See {@link #parseMathExpression(String, Context)} for an
+     * explanation of the syntax.
+     */
+    public static final Pattern EXPRESSION_PATTERN = Pattern.compile("[0-9.,  +\\-*/^()eEkKmMgGbBtT%]*");
+    // Character   to support French locale thousands separator.
```

Comment:
On some environments like GitHub, editor doesn't show NNBSP as visible symbol, so using literal text `NNBSP` would be better.

Reply:
Changed comment, thanks!